src/mds/Capability.cc

```diff
@@ -196,7 +196,8 @@ int Capability::confirm_receipt(ceph_seq_t seq, unsigned caps) {
       dout(10) << "revocation is not totally finished yet on " << *in
                << ", the session " << *session << dendl;
       _revokes.emplace_back(_pending, last_sent, last_issue);
-      calc_issued();
+      if (!is_notable())
```

Comment:
So, the issue primarily is resetting `_issued` with `_pending`, yes?

Could you explain (and mention in the commit message) how this change fixes it as compared to updating _`issued` with only those caps which the client has released.

Reply:
> So, the issue primarily is resetting `_issued` with `_pending`, yes?
> 

Yeah.

> Could you explain (and mention in the commit message) how this change fixes it as compared to updating _`issued` with only those caps which the client has released.

Before receiving the `cap update` request the inode was trying to revoke `Fxr` caps, please see `caps={4941=pAsLsXsFcb/pAsLsXsFxcrb/Fcr@18}`.

While the `cap update` request only released the `Fx` caps, please see `client_caps(update ino 0x3000000025b 1 seq 18 caps=pAsLsXsFcrb`, and still held the `Fr`, and then `confirm_receipt()` would `_issued = pAsLsXsFcrb` and `_pending = pAsLsXsFcb & pAsLsXsFcrb` == `pAsLsXsFcb`

```
184 int Capability::confirm_receipt(ceph_seq_t seq, unsigned caps) {
185   int was_revoking = (_issued & ~_pending);
186   if (seq == last_sent) {
187     _revokes.clear();
188     _issued = caps;                                                                                                                                                                                                                
189     // don't add bits
190     _pending &= caps;
191 
192     // if the revoking is not totally finished just add the
193     // new revoking caps back.
194     if (was_revoking && revoking()) {
195       CInode *in = get_inode();
196       dout(10) << "revocation is not totally finished yet on " << *in
197                << ", the session " << *session << dendl;
198       _revokes.emplace_back(_pending, last_sent, last_issue);
199       calc_issued();
200     }
201   } else {
```
Then in `Line#198` it would add `pAsLsXsFcb` back to revokes list.

While later in `calc_issued()`:

```
349   void calc_issued() {                                                                                                                                                                                                             
350     _issued = _pending;
351     for (const auto &r : _revokes) {    
352       _issued |= r.before;
353     }
354   }
```
The `_issued` would be `_pending | revokes` == `pAsLsXsFcb | pAsLsXsFcb` == `pAsLsXsFcb`. Which would exactly equal to `_pending`.

This was why even the inode need to revoke the `Fr` caps, the logs would be `caps={4941=pAsLsXsFcb/Fcr@18}` after the `cap update` request.

Reply:
> > So, the issue primarily is resetting `_issued` with `_pending`, yes?
> 
> Yeah.
> 
> > Could you explain (and mention in the commit message) how this change fixes it as compared to updating _`issued` with only those caps which the client has released.
> 
> Before receiving the `cap update` request the inode was trying to revoke `Fxr` caps, please see `caps={4941=pAsLsXsFcb/pAsLsXsFxcrb/Fcr@18}`.
 
Right, looks fine till now: `_pending = pAsLsXsFcb`, `_issued = pAsLsXsFxcrb` and `_wanted = Fcr`

which gets interpreted by the client as revoke `Fxr`.

> 
> While the `cap update` request only released the `Fx` caps, please see `client_caps(update ino 0x3000000025b 1 seq 18 caps=pAsLsXsFcrb`, and still held the `Fr`, and then `confirm_receipt()` would `_issued = pAsLsXsFcrb` and `_pending = pAsLsXsFcb & pAsLsXsFcrb` == `pAsLsXsFcb`

The above is fine from code POV.

But, a slightly different question: the `caps=pAsLsXsFcrb` which is part of the MClientCaps update message from the client is actually `implemented` caps rather than `issued`. This change does not touch that, so its probably doing the right thing, but I'm not sure why its done this way...

> 
> ```
> 184 int Capability::confirm_receipt(ceph_seq_t seq, unsigned caps) {
> 185   int was_revoking = (_issued & ~_pending);
> 186   if (seq == last_sent) {
> 187     _revokes.clear();
> 188     _issued = caps;                                                                                                                                                                                                                
> 189     // don't add bits
> 190     _pending &= caps;
> 191 
> 192     // if the revoking is not totally finished just add the
> 193     // new revoking caps back.
> 194     if (was_revoking && revoking()) {
> 195       CInode *in = get_inode();
> 196       dout(10) << "revocation is not totally finished yet on " << *in
> 197                << ", the session " << *session << dendl;
> 198       _revokes.emplace_back(_pending, last_sent, last_issue);
> 199       calc_issued();
> 200     }
> 201   } else {
> ```
> 
> Then in `Line#198` it would add `pAsLsXsFcb` back to revokes list.

Yep - does the right thing (since not all revokes have been released).

> 
> While later in `calc_issued()`:
> 
> ```
> 349   void calc_issued() {                                                                                                                                                                                                             
> 350     _issued = _pending;
> 351     for (const auto &r : _revokes) {    
> 352       _issued |= r.before;
> 353     }
> 354   }
> ```
> 
> The `_issued` would be `_pending | revokes` == `pAsLsXsFcb | pAsLsXsFcb` == `pAsLsXsFcb`. Which would exactly equal to `_pending`.

Right. I understand the issue with blindly updating `_issued` with `_pending`.

> 
> This was why even the inode need to revoke the `Fr` caps, the logs would be `caps={4941=pAsLsXsFcb/Fcr@18}` after the `cap update` request.

Sure. My question in https://github.com/ceph/ceph/pull/52176#discussion_r1244727099 was related to marking the cap as NOTABLE, which pushes the cap object to the front of the xlist. How this fixes the issue needs to be explained in the commit message.

Reply:
> > > So, the issue primarily is resetting `_issued` with `_pending`, yes?
> > 
> > 
> > Yeah.
> > > Could you explain (and mention in the commit message) how this change fixes it as compared to updating _`issued` with only those caps which the client has released.
> > 
> > 
> > Before receiving the `cap update` request the inode was trying to revoke `Fxr` caps, please see `caps={4941=pAsLsXsFcb/pAsLsXsFxcrb/Fcr@18}`.
> 
> Right, looks fine till now: `_pending = pAsLsXsFcb`, `_issued = pAsLsXsFxcrb` and `_wanted = Fcr`
> 
> which gets interpreted by the client as revoke `Fxr`.
> 
> > While the `cap update` request only released the `Fx` caps, please see `client_caps(update ino 0x3000000025b 1 seq 18 caps=pAsLsXsFcrb`, and still held the `Fr`, and then `confirm_receipt()` would `_issued = pAsLsXsFcrb` and `_pending = pAsLsXsFcb & pAsLsXsFcrb` == `pAsLsXsFcb`
> 
> The above is fine from code POV.
> 
> But, a slightly different question: the `caps=pAsLsXsFcrb` which is part of the MClientCaps update message from the client is actually `implemented` caps rather than `issued`. This change does not touch that, so its probably doing the right thing, but I'm not sure why its done this way...
Sorry, didn't get it here. Touch what ?

The `implemented` just only a client side concept, which will be `mds allowed caps` | `being used caps`. So for the cap-update msg here in mds side it should be `issued`.

Actually the `pending' should be `allowed` caps by mds. Right ?

IMO the mds is doing correctly here.
> 
> > ```
[...]
> > The `_issued` would be `_pending | revokes` == `pAsLsXsFcb | pAsLsXsFcb` == `pAsLsXsFcb`. Which would exactly equal to `_pending`.
> 
> Right. I understand the issue with blindly updating `_issued` with `_pending`.
> 
> > This was why even the inode need to revoke the `Fr` caps, the logs would be `caps={4941=pAsLsXsFcb/Fcr@18}` after the `cap update` request.
> 
> Sure. My question in [#52176 (comment)](https://github.com/ceph/ceph/pull/52176#discussion_r1244727099) was related to marking the cap as NOTABLE, which pushes the cap object to the front of the xlist. How this fixes the issue needs to be explained in the commit message.

Okay, I will update the commit comment.
Thanks.

Reply:
> > > > So, the issue primarily is resetting `_issued` with `_pending`, yes?
> > > 
> > > 
> > > Yeah.
> > > > Could you explain (and mention in the commit message) how this change fixes it as compared to updating _`issued` with only those caps which the client has released.
> > > 
> > > 
> > > Before receiving the `cap update` request the inode was trying to revoke `Fxr` caps, please see `caps={4941=pAsLsXsFcb/pAsLsXsFxcrb/Fcr@18}`.
> > 
> > 
> > Right, looks fine till now: `_pending = pAsLsXsFcb`, `_issued = pAsLsXsFxcrb` and `_wanted = Fcr`
> > which gets interpreted by the client as revoke `Fxr`.
> > > While the `cap update` request only released the `Fx` caps, please see `client_caps(update ino 0x3000000025b 1 seq 18 caps=pAsLsXsFcrb`, and still held the `Fr`, and then `confirm_receipt()` would `_issued = pAsLsXsFcrb` and `_pending = pAsLsXsFcb & pAsLsXsFcrb` == `pAsLsXsFcb`
> > 
> > 
> > The above is fine from code POV.
> > But, a slightly different question: the `caps=pAsLsXsFcrb` which is part of the MClientCaps update message from the client is actually `implemented` caps rather than `issued`. This change does not touch that, so its probably doing the right thing, but I'm not sure why its done this way...
> 
> Sorry, didn't get it here. Touch what ?

I meant this change does not modify any of that part (client sending `implemented` instead of `issued`).

Reply:
> The `implemented` just only a client side concept, which will be `mds allowed caps` | `being used caps`. So for the cap-update msg here in mds side it should be `issued`. And the mds `issued` should be the caps are still being `implemented` by client.
> 
> Actually the `pending` should be `allowed` caps by mds.
> 
> Right ?

mds(issued) == client(implemented)
mds(pending) == client(issued)

> 
> IMO the mds is doing correctly here.

Right.

Reply:
Yeah.